### PR TITLE
[optimize] Remember connected gamepad count

### PR DESCRIPTION
**Pull request by PatriceJiang**
Re: https://github.com/cocos/cocos-engine/issues/15619

Prevent query Gamepad events when no devices connected!

### Changelog

* Optimize gamepad event query procedure

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: pal/input/web/gamepad-input.ts [hunk 1]
```diff
@@ -225,42 +227,52 @@ export class GamepadInputDevice {
...
+    private static _totalGamepadCnt = 0;
+    private static _updateGamepadCnt (): void {
+        GamepadInputDevice._totalGamepadCnt = GamepadInputDevice._cachedWebGamepads.reduce((total, gamepad) => (gamepad ? total + 1 : total), 0);
     }
 
     private static _registerEvent (): void {
-        GamepadInputDevice._ensureDirectorDefined().then(() => {
+        GamepadInputDevice._ensureDirectorDefined(() => {
+            GamepadInputDevice._cachedWebGamepads = GamepadInputDevice._getWebGamePads();
+            GamepadInputDevice._updateGamepadCnt();
```
Comment (minggo) [inline pal/input/web/gamepad-input.ts:248]: Why not just add `_totalGamepadCnt`?

File: pal/input/web/gamepad-input.ts [hunk 2]
```diff
@@ -225,42 +227,52 @@ export class GamepadInputDevice {
...
             legacyCC.director.on(legacyCC.Director.EVENT_BEGIN_FRAME, GamepadInputDevice._scanGamepads);
-        }).catch((e) => {});
+        });
         window.addEventListener('gamepadconnected', (e) => {
             GamepadInputDevice._cachedWebGamepads[e.gamepad.index] = e.gamepad;
+            GamepadInputDevice._updateGamepadCnt();
```
Comment (minggo) [inline pal/input/web/gamepad-input.ts:253]: The same as above.

File: pal/input/web/gamepad-input.ts [hunk 3]
```diff
@@ -225,42 +227,52 @@ export class GamepadInputDevice {
...
             const device = GamepadInputDevice._getOrCreateInputDevice(e.gamepad.index, true);
             GamepadInputDevice._eventTarget.emit(InputEventType.GAMEPAD_CHANGE, new EventGamepad(InputEventType.GAMEPAD_CHANGE, device));
         });
         window.addEventListener('gamepaddisconnected', (e) => {
             GamepadInputDevice._cachedWebGamepads[e.gamepad.index] = null;
+            GamepadInputDevice._updateGamepadCnt();
```
Comment (minggo) [inline pal/input/web/gamepad-input.ts:259]: The same as above.

File: pal/input/web/gamepad-input.ts (hunk 1 above)
Comment (PatriceJiang) [reply pal/input/web/gamepad-input.ts:248]: This is where the gamepads are initialized, and a full calculation is needed.

File: pal/input/web/gamepad-input.ts (hunk 2 above)
Comment (PatriceJiang) [reply pal/input/web/gamepad-input.ts:253]: I think this approach ensures that the value of _totalGamepadCnt is always accurate, even if there are changes to _cachedWebGamepads elsewhere in the code.

Comment (minggo) [reply pal/input/web/gamepad-input.ts:253]: Yep, it is always accurate, but need more calculations. But it is ok as it is not frequently invoked.

**Review by minggo — APPROVED** [review]

File: pal/input/web/gamepad-input.ts [hunk 4]
```diff
@@ -225,42 +227,52 @@ export class GamepadInputDevice {
...
-            GamepadInputDevice._intervalId = setInterval(() => {
-                if (legacyCC.director && legacyCC.Director) {
-                    clearInterval(GamepadInputDevice._intervalId);
-                    GamepadInputDevice._intervalId = -1;
-                    resolve();
-                }
-            }, 50);
-        });
+    private static _ensureDirectorDefined (callback: () => void): void {
+        GamepadInputDevice._intervalId = setInterval(() => {
```
Comment (dumganhar) [inline pal/input/web/gamepad-input.ts:231]: Why to remove the `new Promise` code?

Comment (PatriceJiang) [reply pal/input/web/gamepad-input.ts:231]: eslint will report an error, stating that the promise is not awaited. In this specific case, the promise is unnecessary.

Comment (dumganhar) [reply pal/input/web/gamepad-input.ts:231]: Got it.

**Review by dumganhar — APPROVED** [review]

**Review by PPpro — APPROVED** [review]

**Pull request merged**